components/net/http_cache.rs

```diff
@@ -0,0 +1,488 @@
+fn get_expiry_adjustment_from_request_headers(request: &Request, expires: Duration) -> Duration {
+    if let Some(directive_data) = request.headers.get_raw("cache-control") {
+        let directives_string = String::from_utf8(directive_data[0].to_vec()).unwrap();
+        let directives: Vec<&str> = directives_string.split(",").collect();
+        for directive in directives {
+            let directive_info: Vec<&str> = directive.split("=").collect();
+            match directive_info[0] {
+                "max-stale" => {
+                    let seconds = String::from_str(directive_info[1]).unwrap();
+                    return expires + Duration::seconds(seconds.parse::<i64>().unwrap());
+                },
+                "max-age" => {
+                    let seconds = String::from_str(directive_info[1]).unwrap();
+                    let max_age = Duration::seconds(seconds.parse::<i64>().unwrap());
+                    if expires > max_age {
+                        return Duration::min_value();
+                    }
+                    return expires - max_age;
+                },
+                "min-fresh" => {
+                    let seconds = String::from_str(directive_info[1]).unwrap();
+                    let min_fresh = Duration::seconds(seconds.parse::<i64>().unwrap());
+                    if expires < min_fresh {
+                        return Duration::min_value();
+                    }
+                    return expires - min_fresh;
+                },
+                "no-cache" | "no-store" => return Duration::min_value(),
+                _ => {}
+            }
+        }
+    }
+    expires
+}
+
+/// Create a CachedResponse from a request and a CachedResource.
+fn create_cached_response(request: &Request, cached_resource: &CachedResource)
+    -> CachedResponse {
+    let mut response = Response::new(cached_resource.metadata.final_url.clone());
+    response.headers = cached_resource.metadata.headers.lock().unwrap().clone();
+    response.body = cached_resource.body.clone();
+    response.status = cached_resource.status.clone();
+    response.raw_status = cached_resource.raw_status.clone();
+    response.url_list = cached_resource.url_list.clone();
+    let expires = *cached_resource.expires.lock().unwrap();
+    let adjusted_expires = get_expiry_adjustment_from_request_headers(request, expires);
+    let now = Duration::seconds(time::now().to_timespec().sec);
+    let last_validated = Duration::seconds(cached_resource.last_validated.to_timespec().sec);
+    let time_since_validated = now - last_validated;
+    // TODO: take must-revalidate into account https://tools.ietf.org/html/rfc7234#section-5.2.2.1
+    // TODO: since this cache is shared, taking proxy-revalidate into account
+    // https://tools.ietf.org/html/rfc7234#section-5.2.2.7
+    let has_expired = (adjusted_expires < time_since_validated)
+        | (adjusted_expires == time_since_validated);
```

Comment:
Let's use ||, and move the operator to the previous line.

Reply:
Ok